cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetTypeService.java

```diff
@@ -0,0 +1,417 @@
+    File tempDir = new File(new File(namespaceBase, datasetModuleId.getNamespace()),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    if (!DirUtils.mkdirs(tempDir)) {
+      throw new IOException("Could not create temporary directory at: " + tempDir);
+    }
+
+    return new AbstractBodyConsumer(File.createTempFile("dataset-", ".jar", tempDir)) {
+      @Override
+      protected void onFinish(HttpResponder responder, File uploadedFile) throws Exception {
+        if (className == null) {
+          // We have to delay until body upload is completed due to the fact that not all client is
+          // requesting with "Expect: 100-continue" header and the client library we have cannot handle
+          // connection close, and yet be able to read response reliably.
+          // In longer term we should fix the client, as well as the netty-http server. However, since
+          // this handler will be gone in near future, it's ok to have this workaround.
+          responder.sendString(HttpResponseStatus.BAD_REQUEST, "Required header 'class-name' is absent.");
+          return;
+        }
+
+        LOG.info("Adding module {}, class name: {}", datasetModuleId, className);
+
+        String dataFabricDir = cConf.get(Constants.Dataset.Manager.OUTPUT_DIR);
+        String moduleName = datasetModuleId.getModule();
+        Location archiveDir = namespaceHomeLocation.append(dataFabricDir).append(moduleName)
+          .append(Constants.ARCHIVE_DIR);
+        String archiveName = moduleName + ".jar";
+        Location archive = archiveDir.append(archiveName);
+
+        // Copy uploaded content to a temporary location
+        Location tmpLocation = archive.getTempFile(".tmp");
+        try {
+          Locations.mkdirsIfNotExists(archiveDir);
+
+          LOG.debug("Copy from {} to {}", uploadedFile, tmpLocation);
+          Files.copy(uploadedFile, Locations.newOutputSupplier(tmpLocation));
+
+          // Finally, move archive to final location
+          LOG.debug("Storing module {} jar at {}", datasetModuleId, archive);
+          if (tmpLocation.renameTo(archive) == null) {
+            throw new IOException(String.format("Could not move archive from location: %s, to location: %s",
+                                                tmpLocation, archive));
+          }
+
+          typeManager.addModule(datasetModuleId.toId(), className, archive, forceUpdate);
+          // todo: response with DatasetModuleMeta of just added module (and log this info)
+          // Ideally this should have been done before, but we cannot grant privileges on types until they've been
+          // added to the type MDS. First revoke any orphaned privileges for types left behind by past failed revokes
+          revokeAllPrivilegesOnModule(datasetModuleId);
+          grantAllPrivilegesOnModule(datasetModuleId, principal);
+          LOG.info("Added module {}", datasetModuleId);
```

Comment:
This is a success message upon dataset module deployment. Should we keep it? It is unlikely that this will be a frequent operation in any cluster - Users will have to create a non-system, custom dataset module and deploy it to see this log.